### PR TITLE
Better support for streaming datasets; multidataset weighting / round robin

### PR DESCRIPTION
**Pull request by djsaunde**
# Description

Streaming datasets should not be limited to pretraining only. This PR changes that, and also adds support for different sampling strategies (round robin and sampling according to weights). Also, preprocessing iterable datasets has been disallowed by this PR.

## Motivation and Context

Some folks have been asking for better dataset streaming support. This is part of a larger effort to refactor / improve the logic around how we handle datasets / dataloading.

## How has this been tested?

- A few tests (need more)
- A few manual runs

## Summary by CodeRabbit

* **New Features**
  * Streaming dataset support with per-split handling, sharding, on-demand loading, and dataset mixing strategies (round_robin, weighted, random, concatenate).

* **Bug Fixes / Validation**
  * Stronger config validation for streaming, max_steps/num_epochs conflicts, and mixing-weights consistency.

* **Documentation**
  * Enhanced schema descriptions for epochs, streaming, and mixing options.

* **Refactor**
  * Removed legacy dataset-packing/streaming preprocess path and simplified preprocessing flow.

* **Tests**
  * Added unit and end-to-end tests covering streaming and dataset-mixing behaviors.

* **Chores**
  * Added config fields: streaming, dataset_mixing_strategy, mixing_weights; removed preprocess_iterable.

### CODE REVIEW

**Comment by djsaunde** [comment]
Currently seeing an [error](https://gist.github.com/djsaunde/5e1d86046275654f2dd7f174af15a567) when using eval streaming datasets, will need to debug

File: src/axolotl/datasets.py [hunk 1]
```diff
@@ -79,16 +98,15 @@ def wrap_dataset_for_tokenized_prompt(
...
             prompt_tokenizer.tokenize_prompt,
-            remove_columns=features,
             **map_kwargs,
         )
     return TokenizedPromptDataset(prompt_tokenizer, dataset, **kwargs)
 
 
-# TODO this isn't the best since it can't interleave datasets
+# TODO: this isn't the best since it can't interleave datasets.
+# NOTE: this is only used in a test. Can it be deleted?
```
Comment (winglian) [inline src/axolotl/datasets.py]: yeah, let's delete this. This was from before we added  Multipack Sampler

File: src/axolotl/cli/args.py [hunk 2]
```diff
@@ -13,12 +13,6 @@ class PreprocessCliArgs:
     debug_num_examples: int = field(default=1)
     prompter: Optional[str] = field(default=None)
     download: Optional[bool] = field(default=True)
-    iterable: Optional[bool] = field(
```
Comment (salmanmohammadi) [inline src/axolotl/cli/args.py:16]: No deprecation strategy?

**Comment by salmanmohammadi** [comment]
What did you find in your manual testing?

File: src/axolotl/datasets.py [hunk 3]
```diff
@@ -1,18 +1,21 @@
-"""Module containing Dataset functionality"""
+"""
+Module containing dataset functionality.
+
+We want this to be a wrapper for an existing dataset that we have loaded. Lets use the
+concept of middlewares to wrap each dataset, for example:
+ConstantLengthDataset(ShuffledDataset([TokenizedPromptDataset(alpaca_dataset)])).
+Let's check to ensure we don't truncate an item in the middle. We'll use the collators
+later on to pad the datasets.
+"""
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/datasets.py]: good robot
@djsaunde

File: src/axolotl/datasets.py [hunk 4]
```diff
@@ -79,140 +97,8 @@ def wrap_dataset_for_tokenized_prompt(
         map_kwargs = {}
         if prompt_tokenizer.supports_batched:
             map_kwargs["batched"] = True
-        features = list(dataset.features.keys())
         return dataset.map(
             prompt_tokenizer.tokenize_prompt,
-            remove_columns=features,
```
Comment (winglian) [inline src/axolotl/datasets.py:85]: iirc, for datasets that have other columns that get mapped into new columns, not removing them leads to errors with passing those columns into the model forward.

File: src/axolotl/cli/args.py (hunk 2 above)
Comment (djsaunde) [reply src/axolotl/cli/args.py:16]: I was thinking that probably no one uses this, but to be safe I should have a deprecation strategy.

**Comment by djsaunde** [comment]
> What did you find in your manual testing?

@SalmanMohammadi Basic tests worked with no noticeable extra data time. I should get benchmark numbers (esp for more complicated situations; e.g., multidataset, multiprocess) before merge, though.

File: src/axolotl/utils/data/shared.py [hunk 5]
```diff
@@ -562,6 +565,45 @@ def merge_datasets(datasets: list[Dataset], cfg: DictDefault) -> Dataset:
...
-        LOG.debug("Not shuffling merged datasets.")
+        if isinstance(merged_dataset, IterableDataset):
+            LOG.debug("Skipping shuffle for streaming datasets.")
+        else:
+            LOG.debug("Not shuffling merged datasets.")
 
     return merged_dataset
+
+
+def _merge_datasets_with_strategy(
```
Comment (winglian) [inline src/axolotl/utils/data/shared.py:576]: 🔥

File: src/axolotl/utils/trainer.py [hunk 6]
```diff
@@ -23,6 +22,65 @@
...
+                    if keep:
+                        yield batch[i]
+        else:
+            # For non-batched filtering, apply filter to each sample individually
+            for sample in dataset:
+                if filter_fn(sample):
+                    yield sample
+
+    # Create new IterableDataset from the filtered generator
+    filtered_dataset = IterableDataset.from_generator(filtered_generator)
```
Comment (winglian) [inline src/axolotl/utils/trainer.py:74]:
```suggestion
    filtered_dataset = IterableDataset.from_generator(filtered_generator, gen_kwargs={"batch_size": 1000})
```
`batch_size` as a kwarg feels better than a constant

**Review by winglian — APPROVED** [review]
Just a couple of nits. lgtm. Were you able to do any loss curve and step count comparisons to standard SFT w/o streaming?

**Comment by djsaunde** [comment]
> Just a couple of nits. lgtm. Were you able to do any loss curve and step count comparisons to standard SFT w/o streaming?

Yes, everything looked okay except for sample packing + streaming, will need to fix.

**Comment by winglian** [comment]
> Yes, everything looked okay except for sample packing + streaming, will need to fix.

packing and streaming is something that I struggled with due to IterableDatasets and BatchSampler within accelerate iirc.
Screenshot 2025-08-23 at 12 50 27 PM

**Comment by djsaunde** [comment]
Closing in favor of #3101.
